components/places/sql/create_sync_temp_tables.sql

```diff
@@ -0,0 +1,134 @@
+
+-- A view of all synced items. We use triggers on this view to update local
+-- items. Note that we can't just `REPLACE INTO moz_bookmarks`, because
+-- `REPLACE` doesn't fire the shared `AFTER DELETE` triggers that we use to
+-- maintain schema coherency.
+CREATE TEMP VIEW itemsToMerge(localId, localGuid, remoteId, remoteGuid,
+                              mergedGuid, useRemote, shouldUpload, newLevel,
+                              newType,
+                              newDateAddedMicroseconds,
+                              newTitle, oldPlaceId, newPlaceId,
+                              newKeyword) AS
+SELECT b.id, b.guid, v.id, v.guid,
+       r.mergedGuid, r.useRemote, r.shouldUpload, r.level,
+       (CASE WHEN v.kind IN (
+                    1, -- SyncedBookmarkKind::Bookmark
+                    2 -- SyncedBookmarkKind::Query
+                  ) THEN 1 -- BookmarkType::Bookmark
+             WHEN v.kind IN (
+                    3, -- SyncedBookmarkKind::Folder
+                    4 -- SyncedBookmarkKind::Livemark
+                  ) THEN 2 -- BookmarkType::Folder
+             ELSE 3 -- BookmarkType::Separator
+             END),
+       /* Take the older creation date. "b.dateAdded" is in microseconds;
+          "v.dateAdded" is in milliseconds. */
+       (CASE WHEN b.dateAdded / 1000 < v.dateAdded THEN b.dateAdded
+             ELSE v.dateAdded * 1000 END),
+       v.title, h.id, (SELECT n.id FROM moz_places n
+                       WHERE n.url_hash = u.hash AND
+                             n.url = u.url),
+       v.keyword
+FROM mergedTree r
+LEFT JOIN moz_bookmarks_synced v ON v.guid = r.remoteGuid
+LEFT JOIN moz_bookmarks b ON b.guid = r.localGuid
+LEFT JOIN moz_places h ON h.id = b.fk
+LEFT JOIN urls u ON u.id = v.urlId
+WHERE r.mergedGuid <> "root________";
+
+-- A view of the structure states for all items in the merged tree.
+CREATE TEMP VIEW structureToMerge(localId, oldParentId, newParentId,
+                                  oldPosition, newPosition, newLevel) AS
+SELECT b.id, b.parent, p.id, b.position, r.position, r.level
+FROM moz_bookmarks b
+JOIN mergedTree r ON r.mergedGuid = b.guid
+JOIN moz_bookmarks p ON p.guid = r.mergedParentGuid
+/* Don't reposition roots, since we never upload the Places root, and our
+   merged tree doesn't have a tags root. */
+WHERE "root________" NOT IN (r.mergedGuid, r.mergedParentGuid);
+
+-- Stores local IDs for items to upload even if they're not flagged as changed
+-- in Places. These are "weak" because we won't try to reupload the item on
+-- the next sync if the upload is interrupted or fails.
+CREATE TEMP TABLE idsToWeaklyUpload(
+  id INTEGER PRIMARY KEY
+);
+
+-- Stores local IDs for items to reupload. Removing an
+-- ID from this table bumps its local change counter, so, unlike weak uploads,
+-- we *will* reupload the item on the next sync if the current sync fails.
+-- This is used to ensure that all bookmarks with the same URL have the same keyword (bug 1328737).
```

Comment:
I think we can remove this complexity for keywords. Since we're not exposing any UI to edit them (IOW, keyword changes will only be remote, not local), we can just round-trip them, and store them with the bookmark instead of maintaining keyword-URL associations.